### PR TITLE
[MoveOnlyAddressChecker] Fixed two use-before-def handlings.

### PR DESCRIPTION
**Pull request by nate-chandler**
FieldSensitivePL's handling of uses-before-defs was exposed by https://github.com/apple/swift/pull/66826 to be incorrect.  The fix for that bug in turn exposed that the handling of uses-before-defs in lifetime maximization's original live block discovery was incorrect.

rdar://111118843&111221183

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

File: lib/SILOptimizer/Mandatory/MoveOnlyAddressCheckerUtils.cpp [hunk 1]
```diff
@@ -2971,15 +2974,15 @@ void ExtendUnconsumedLiveness::runOnField(
     }
 
     // Walk backwards from consuming blocks.
+    BasicBlockBitfield cache(currentDef->getFunction(), 2);
     while (auto *block = worklist.pop()) {
-      if (!originalLiveBlocks.insert(block)) {
+      if (!originalLiveBlocks.insert(block)
+          // Don't walk into the predecessors of blocks which kill liveness.
+          && !isLiveAtBegin(block, element, /*isLiveAtEnd=*/true, destroys,
```
Comment (atrick) [inline lib/SILOptimizer/Mandatory/MoveOnlyAddressCheckerUtils.cpp:2981]: As discussed, we probably don't need to check local liveness for each block on the worklist. So you probably then don't even need the cache,

**Review by atrick — APPROVED** [review]
Awesome! Thank you.

There's one comment here about simplifying original liveness propagation.

File: lib/SILOptimizer/Mandatory/MoveOnlyAddressCheckerUtils.cpp (hunk 1 above)
Comment (nate-chandler) [reply lib/SILOptimizer/Mandatory/MoveOnlyAddressCheckerUtils.cpp]: Fixed.

**Comment by nate-chandler** [comment]
@swift-ci please test

**Pull request merged**